### PR TITLE
feat: NFT to fungible token manipulation

### PR DESCRIPTION
**Pull request by kupermind**
- Initial code and tests.

### CODE REVIEW

File: solidity/nft_token.sol [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+    address position_mint;  // 32
+    uint128 liquidity;      // 16
+    int32 tick_lower_index; // 4
+    int32 tick_upper_index; // 4
+}
+
+@program_id("DPvSokZKEAYqh6jjmGDUyd4L881uAWg4rFVeA6rec2fr")
+contract nft_token {
+    address public orca;
+    address public whirlpool;
```
Comment (DavidMinarsch) [inline solidity/nft_token.sol:15]: What's whirlpool?

File: solidity/nft_token.sol [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+    int32 public constant min_tick_lower_index = -443632;
+    int32 public constant max_tick_lower_index = 443632;
+    //address constant tokenProgram = address"TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA";
+
+    uint64 public numTokenAccounts;
+
+    mapping(address => uint64) public mapTokenAccountLiquidity;
+    address[type(uint32).max] public tokenAccounts;
+
+    @space(10000)
```
Comment (DavidMinarsch) [inline solidity/nft_token.sol:34]: ?

File: solidity/nft_token.sol [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+    address public whirlpool;
+    address public pdaProgram;
+    address public mintErc20;
+    address public pdaERC20Account;
+    // TODO: Change for the latter one in production
+    uint64 public pdaHeader = 0x55bdfe33;//0xd0f7407ae48fbcaa;
+    // PDA seed
+    bytes public constant pdaProgramSeed = "pdaProgram";
+    // PDA bump
+    bytes1 public pdaBump;
```
Comment (DavidMinarsch) [inline solidity/nft_token.sol:24]: What is this?

File: solidity/nft_token.sol [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+        // Verify that the bump passed to the constructor matches the bump derived from the seeds and programId
+        if (bump != _bump) {
+            revert("Invalid bump");
+        }
+
+        pdaBump = bump;
+        pdaProgram = pda;
+    }
+
+    function getPositionData(address positionDataAccount, address positionMint) public view returns (Position position) {
```
Comment (DavidMinarsch) [inline solidity/nft_token.sol:61]: Presumably the liquidity pair needs to exist before we can deploy this?

File: solidity/nft_token.sol [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+        // TODO: Do the liquidity check for max(uint64) value as it is provided as uint128 from the LP provider
+        uint64 liquidity = uint64(position.liquidity);
+
+        // Transfer the NFT to the pdaTokenAccount address of this program
+        SplToken.transfer(
+            tx.accounts.fromTokenAccount.key,
+            tx.accounts.pdaTokenAccount.key,
+            tx.accounts.fromWallet.key,
+            1);
+
```
Comment (DavidMinarsch) [inline solidity/nft_token.sol:171]: Not sure if reentrancy is a thing in Solana, still suggest doing checks-effects-interactions pattern here and elsewhere! It's a good habit imo

File: solidity/nft_token.sol [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+    @mutableAccount(mintERC20)
+    @signer(sig)
+    // Transfer with PDA
+    function withdraw(uint64 amount) external {
+        // Check that the token account exists
+        if (mapTokenAccountLiquidity[tx.accounts.pdaTokenAccount.key] == 0) {
+            revert("No liquidity on a provided token account");
+        }
+
+        // Transfer ERC20 tokens to the pdaERC20Account address of this program
```
Comment (DavidMinarsch) [inline solidity/nft_token.sol:202]: Should it not be burned?

Comment (DavidMinarsch) [reply solidity/nft_token.sol:202]: Ah happens below

File: solidity/nft_token.sol [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+            amount);
+
+        // Transfer NFT to the user associated token account
+        SplToken.transfer_pda(
+            tx.accounts.pdaTokenAccount.key,
+            tx.accounts.fromTokenAccount.key,
+            pdaProgram,
+            1,
+            pdaProgramSeed,
+            pdaBump);
```
Comment (DavidMinarsch) [inline solidity/nft_token.sol:216]: Don't we have to account for the potential splitting of the liquidity here?

File: solidity/nft_token.sol (hunk 2 above)
Comment (77ph) [reply solidity/nft_token.sol:34]: if you are creating a PDA account, you're limited to 10kb for the account size.
space is "solang helper" for allocated space for "PDA" data account.
"Alternatively, the data account can be created by the constructor, on chain. When this method is used, some parameters must be specified for the account using annotations. Annotations placed above a constructor can only contain literals or constant expressions, as is the case for first @seed and @space"
This is a feature of Solana and such a bridge is made by Solang developers to use a language similar to Solidity along with the concept of Solana accounts.
If you don’t do this or don’t do it off-chain, then when you try to write something during execution program  to "program" data account (in EVM term "write to storage"), you will get an error.

File: solidity/nft_token.sol (hunk 3 above)
Comment (77ph) [reply solidity/nft_token.sol:24]: https://www.brianfriel.xyz/understanding-program-derived-addresses/
"fudge factor" (int8 = 255 .. 0) for "push" potencial PDA off the ed25519 elliptic curve
https://solanacookbook.com/core-concepts/pdas.html#generating-pdas
"When building with PDAs, it is common to store the bump seed in the account data itself. This allows developers to easily validate a PDA without having to pass in the bump as an instruction argument."
We simply follow examples and practices.

File: solidity/nft_token.sol (hunk 5 above)
Comment (77ph) [reply solidity/nft_token.sol:171]: Yes. for solana reentrancy issue very rare.
1) https://medium.com/coinmonks/on-smart-contracts-why-solana-is-more-secure-18626b09929b
https://github.com/solana-developers/solana-cookbook/issues/163
- CPI is limited, we can endlessly call one contract from another into the EVM until the gas runs out. in Solana CPI - 4 max.
- there is no concept of a token as a program. there are no arbitrary program (contract) made by users. we call a limited number of trusted programs.
- EVM's callback pattern (A -> B -> A) cannot be implemented. (c) Based on what I understand about CPI, this seems correct.

In terms of security, we need to focus on verifying accounts.
Example here:
https://medium.com/coinmonks/cashioapp-attack-whats-the-vulnerability-and-how-soteria-detects-it-2e96b9c6d1d3
We tried to implement checks.

File: solidity/nft_token.sol (hunk 1 above)
Comment (kupermind) [reply solidity/nft_token.sol:15]: It's an account that holds the overall pool position with the full liquidity.

File: solidity/nft_token.sol (hunk 4 above)
Comment (kupermind) [reply solidity/nft_token.sol:61]: Yes, however we will revise this check in a next iteration.

File: solidity/nft_token.sol (hunk 7 above)
Comment (kupermind) [reply solidity/nft_token.sol:216]: This part is not yet complete, but we are certain now that we don't need to do that, as decreaseLiquidity() function removes the specified liquidity from a specific position. This way we don't need to completely liquidate a full position and create another one from remainders later.

File: solidity/library/spl_token.sol [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+		AccountMeta[3] metas = [
+			AccountMeta({pubkey: account, is_writable: true, is_signer: false}),
+			AccountMeta({pubkey: mint, is_writable: true, is_signer: false}),
+			AccountMeta({pubkey: owner, is_writable: false, is_signer: true})
+		];
+
+		tokenProgramId.call{accounts: metas, seeds: [[seed, bump]]}(instr);
+	}
+
+	/// Approve an amount to a delegate. This transaction should be signed by the owner
```
Comment (mariapiamo) [inline solidity/library/spl_token.sol:199]:
```suggestion
	/// Approve an amount to a delegate. This transaction should be signed by the owner of the account
```

File: solidity/nft_token.sol [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+    address public mintErc20;
+    address public pdaERC20Account;
+    // TODO: Change for the latter one in production
+    uint64 public pdaHeader = 0x55bdfe33;//0xd0f7407ae48fbcaa;
+    // PDA seed
+    bytes public constant pdaProgramSeed = "pdaProgram";
+    // PDA bump
+    bytes1 public pdaBump;
+    int32 public constant min_tick_lower_index = -443632;
+    int32 public constant max_tick_lower_index = 443632;
```
Comment (mariapiamo) [inline solidity/nft_token.sol:26]: to adjust with the decimals

File: solidity/library/spl_token.sol (hunk 8 above)
Comment (kupermind) [reply solidity/library/spl_token.sol:199]: It's the copy-paste of the code ;)

**Pull request merged**